### PR TITLE
Construct UOps patterns using UPat

### PR DESCRIPTION
**Pull request by alecchendev**
Followup on #4791. While doing this, I took the chance to use positional arguments wherever it wasn't too hard, let me know if I should reverse these.

### CODE REVIEW

**Comment by geohot** [comment]
Excellent use of process replay. @Qazalin I'm assuming by the long time CI took that it passed? Why is it so slow, shouldn't it just compare a hash?

File: tinygrad/codegen/uops.py [hunk 1]
```diff
@@ -77,30 +77,19 @@ class UPat:
...
-  if isinstance(pat.arg, set) and uop.arg not in pat.arg: return False
-  elif pat.arg is not None and uop.arg != pat.arg: return False
-  if isinstance(pat.dtype, set) and uop.dtype not in pat.dtype: return False
-  if isinstance(pat.dtype, DType) and uop.dtype != pat.dtype: return False
-  if isinstance(pat.uop, set) and uop.uop not in pat.uop: return False
-  if isinstance(pat.uop, UOps) and uop.uop != pat.uop: return False
+  if pat.name is not None:
+    if pat.name in store and store[pat.name] != uop: return False
+    store[pat.name] = uop
+  if pat.arg is not None and uop.arg != pat.arg: return False
```
Comment (geohot) [inline tinygrad/codegen/uops.py:84]: Err, arg can be a set now after the PTX shift addition? Why was this not caught by process replay?

**Comment by Qazalin** [comment]
@geohot yes they passed! https://github.com/tinygrad/tinygrad/actions/runs/9357955542/job/25758850270
I'm pushing to update_benchmark to do the same.

**Comment by geohot** [comment]
I don't understand how the PTX one passed...from what I read it no longer correctly handles "arg" as a set.

**Comment by Qazalin** [comment]
process replay runs ASSERT_COMPILE=1 against a cache built up from master, maybe no kernels changed for the limited tests we run on ptx CI? https://github.com/tinygrad/tinygrad/blob/master/.github/workflows/test.yml#L487

File: tinygrad/codegen/uops.py (hunk 1 above)
Comment (alecchendev) [reply tinygrad/codegen/uops.py:84]: Oops, must've accidentally skipped over this when rebasing, will also add a regression test for this

Comment (alecchendev) [reply tinygrad/codegen/uops.py]: fwiw the previous implementation did not pass the regression test i just added

**Comment by alecchendev** [comment]
IIUC it looks like the logic added to match an arg set in #4728 would never match the positive case where the arg is in the set, so the patterns added in that PR never got used. With the latest commit here this is fixed, and now the ptx test is failing

**Comment by Qazalin** [comment]
@alecchendev hmm, can you prove this with a separate PR for a regression test?

**Comment by geohot** [comment]
@SzymonOzog Can you add a regression test confirming shift is actually selected on master?

**Comment by alecchendev** [comment]
> @alecchendev hmm, can you prove this with a separate PR for a regression test? maybe by adapting test_arg_set to the master PatternMatcher

#4827

**Comment by geohot** [comment]
Congrats, assuming this passes master CI, bounty is yours!

Can pay via PayPal or USDC, e-mail george@tinygrad.org to claim

**Pull request merged**